CFG pattern dispose with args

chsienki opened the pull request:
Correctly report the arguments for pattern disposal when there are default or params arguments

chsienki commented:
Follow up from #49481 

@dotnet/roslyn-compiler for review please 

//cc @AlekseyTs and @333fred

RikkiGibson commented on src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:
```diff
@@ -2337,5 +2341,44 @@ private IInstanceReferenceOperation CreateCollectionValuePlaceholderOperation(Bo
+                out var defaultArguments,
+                expanded,
+                enableCallerInfo: true,
+                new DiagnosticBag());
```
should we use a pooled diagnostic bag (even if the diagnostics will be discarded?)

AlekseyTs commented on src/Compilers/Core/Portable/Operations/OperationInterfaces.xml:
```diff
@@ -2944,6 +2949,13 @@
         <summary>The method that will be invoked to dispose the declared instances when pattern based disposal is used.</summary>
       </Comments>
     </Property>
+    <Property Name="DisposeArguments" Type="ImmutableArray&lt;IArgumentOperation&gt;" Internal="true">
```
>DisposeArguments [](start = 20, length = 16)

I would expect the two test IOperation visitors that we have to be adjusted appropriately to verify the value in this property. Same for the property above. It probably makes sense to ignore an empty array in the visitor that dumps the tree in order to avoid changes in baselines for tests that do not have arguments.

AlekseyTs commented on src/Compilers/Core/Portable/Operations/OperationInterfaces.xml:
```diff
@@ -490,6 +490,11 @@
         <summary>The method that will be invoked to dispose the <see cref="Resources" /> when pattern based disposal is used.</summary>
       </Comments>
     </Property>
+    <Property Name="DisposeArguments" Type="ImmutableArray&lt;IArgumentOperation&gt;" Internal="true">
```
>DisposeArguments [](start = 20, length = 16)

Same comment as below

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -3936,15 +3936,17 @@ private void AddDisposingFinally(IOperation resource, bool requiresRuntimeConver
             IOperation? tryDispose(IOperation value)
             {
-                Debug.Assert(disposeMethod is object || value.Type!.Equals(iDisposable));
+                Debug.Assert((disposeMethod is object && !disposeArguments.IsDefault) || value.Type!.Equals(iDisposable));
```
>value.Type!.Equals(iDisposable) [](start = 89, length = 31)

It feels like we should be able to assert something about ```disposeArguments``` on this side as well.

AlekseyTs commented:
>                          MakeInvalidOperation(type: null, resource));

I think to future proof the builder we should push/pop a stack frame around this AddStatement call. See how we do this in getEnumerator helper.

---
Refers to: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:3929 in d81e555. [](commit_id = d81e555d5cb10fd2fd7cdc6134481ddc8af79cfd, deletion_comment = False)

chsienki commented:
@AlekseyTs Thanks Aleksey, I think I've addressed all your feedback. In adding the test visits, I realized we didn't have any non-flow tests exercising them, so added a pair of `_DefaultDisposeArguments` that verify the actual IOperations when we have default arguments present. 

I was going to add one to the foreach tests, but it doesn't look like we've ever printed the disposal status (or getEnumerator etc), so I don't think we need to.

AlekseyTs commented on src/Compilers/Test/Core/Compilation/OperationTreeVerifier.cs:
```diff
@@ -469,6 +469,13 @@ public override void VisitUsingDeclaration(IUsingDeclarationOperation operation)
 
             Visit(operation.DeclarationGroup, "DeclarationGroup");
+
+            var disposeMethod = ((UsingDeclarationOperation)operation).DisposeMethod;
```
>disposeMethod [](start = 16, length = 13)

I think it would be good to include this symbol into the output when it is not null.

AlekseyTs commented on src/Compilers/Test/Core/Compilation/OperationTreeVerifier.cs:
```diff
@@ -739,6 +746,13 @@ public override void VisitUsing(IUsingOperation operation)
             Assert.NotEqual(OperationKind.VariableDeclaration, operation.Resources.Kind);
             Assert.NotEqual(OperationKind.VariableDeclarator, operation.Resources.Kind);
+
+            var disposeMethod = ((UsingOperation)operation).DisposeMethod;
```
>disposeMethod [](start = 16, length = 13)

Same comment as for VisitUsingDeclaration

AlekseyTs commented on src/Compilers/Test/Core/Compilation/TestOperationVisitor.cs:
```diff
@@ -424,6 +425,16 @@ public override void VisitUsing(IUsingOperation operation)
+            {
+                foreach (var arg in disposeArgs)
+                {
+                    VisitArgument(arg);
```
>VisitArgument [](start = 20, length = 13)

Should we use VerifySubTree here?

AlekseyTs commented on src/Compilers/Test/Core/Compilation/TestOperationVisitor.cs:
```diff
@@ -1530,6 +1541,16 @@ public override void VisitUsingDeclaration(IUsingDeclarationOperation operation)
+            {
+                foreach (var arg in disposeArgs)
+                {
+                    VisitArgument(arg);
```
>VisitArgument [](start = 20, length = 13)

Should we use VerifySubTree here instead?

AlekseyTs commented on src/Compilers/Core/Portable/Generated/Operations.Generated.cs:
```diff
@@ -3729,20 +3729,22 @@ public override IEnumerable<IOperation> Children
             Locals = locals;
             IsAsynchronous = isAsynchronous;
             DisposeMethod = disposeMethod;
+            DisposeArguments = SetParentOperation(disposeArguments, this);
```
>DisposeArguments = SetParentOperation(disposeArguments, this); [](start = 12, length = 62)

It looks like this sub-tree is getting different treatment by comparison to similar sub-trees for foreach. I think we should align the treatment and follow what we do for foreach.

AlekseyTs commented on src/Compilers/Core/Portable/Generated/Operations.Generated.cs:
```diff
@@ -6474,16 +6476,18 @@ public override IEnumerable<IOperation> Children
             DeclarationGroup = SetParentOperation(declarationGroup, this);
             IsAsynchronous = isAsynchronous;
             DisposeMethod = disposeMethod;
+            DisposeArguments = SetParentOperation(disposeArguments, this);
```
>DisposeArguments = SetParentOperation(disposeArguments, this); [](start = 12, length = 62)

Same comment as for UsingOperation

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/UsingStatementBinder.cs:
```diff
@@ -156,7 +156,7 @@ internal static BoundStatement BindUsingStatementOrDeclarationFromParts(SyntaxNo
             if (isUsingDeclaration)
             {
-                return new BoundUsingLocalDeclarations(syntax, disposeMethodOpt, iDisposableConversion, awaitOpt, declarationsOpt, hasErrors);
+                return new BoundUsingLocalDeclarations(syntax, disposeMethodOpt, iDisposableConversion, awaitOpt, originalBinder, declarationsOpt, hasErrors);
```
>originalBinder [](start = 114, length = 14)

Could this be an SemanticModel binder (an IncrementalBinder)? We probably don't want to cache it in the tree and should skip to the "regular" binder. Should probably check other places where we cache binders in bound tree.

jaredpar replied:
Looks like a couple of other `BoundNode` store binders: `BoundCallExpression` and `BoundObjectInitializerMember` for example. These end up storing the `Binder` value `this` which is an arbitrary sub-class of `Binder`. 

Can you help me understand why that is safe but storing `originalBinder` is not?

AlekseyTs replied:
> Can you help me understand why that is safe but storing originalBinder is not? 

I said: "Should probably check other places where we cache binders in bound tree." And those are good examples to check. Also, I am not sure if "unsafe" is the right term to use here. It is probably unlikely to find an observable negative effect from caching an IncrementalBinder besides holding onto an object that we expect to be transient and short lived. An IncrementalBinder also "modifies" the state of SemanticModel it is associated with. Depending on the usage of the cached binder, that might never happen, or if it happens, this might be desirable or not desirable. By default, I would assume that any effect on SemanticModel is not desirable. There could be exceptions, of course. 

If the sole purpose of capturing the binder is to build arguments later. I think we simply should build the arguments on the spot and capture them instead.

---
In reply to: [539670556](https://github.com/dotnet/roslyn/pull/49842#discussion_r539670556) [](ancestors = 539670556)

AlekseyTs replied:
> If the sole purpose of capturing the binder is to build arguments later. I think we simply should build the arguments on the spot and capture them instead.

Doing this is probably going to affect other nodes/scenarios, better to deal with as a separate change

---
In reply to: [539703344](https://github.com/dotnet/roslyn/pull/49842#discussion_r539703344) [](ancestors = 539703344,539670556)

AlekseyTs replied:
And if we plan to do that later, then probably no reason to worry about skipping binders here now.

---
In reply to: [539704481](https://github.com/dotnet/roslyn/pull/49842#discussion_r539704481) [](ancestors = 539704481,539703344,539670556)

AlekseyTs replied:
See [https://github.com/dotnet/roslyn/issues/49218](https://github.com/dotnet/roslyn/issues/49218)

---
In reply to: [539706285](https://github.com/dotnet/roslyn/pull/49842#discussion_r539706285) [](ancestors = 539706285,539704481,539703344,539670556)

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/MethodToClassRewriter.cs:
```diff
@@ -193,7 +193,7 @@ public override BoundNode VisitUsingStatement(BoundUsingStatement node)
             BoundStatement body = (BoundStatement)this.Visit(node.Body);
             Conversion disposableConversion = RewriteConversion(node.IDisposableConversion);
-            return node.Update(newLocals, declarationsOpt, expressionOpt, disposableConversion, body, node.AwaitOpt, node.DisposeMethodOpt);
+            return node.Update(newLocals, declarationsOpt, expressionOpt, disposableConversion, body, node.AwaitOpt, node.DisposeMethodOpt, node.Binder);
```
>node.Binder [](start = 140, length = 11)

Can we drop the binder at this point?

AlekseyTs commented on src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:
```diff
@@ -1771,9 +1772,11 @@ private IUsingOperation CreateBoundUsingStatementOperation(BoundUsingStatement b
             ImmutableArray<ILocalSymbol> locals = boundUsingStatement.Locals.GetPublicSymbols();
             bool isAsynchronous = boundUsingStatement.AwaitOpt != null;
             IMethodSymbol? disposeMethod = boundUsingStatement.DisposeMethodOpt.GetPublicSymbol();
+            var receiver = boundUsingStatement.ExpressionOpt ?? boundUsingStatement.DeclarationsOpt?.LocalDeclarations[0].DeclaredTypeOpt;
```
>var receiver = boundUsingStatement.ExpressionOpt ?? boundUsingStatement.DeclarationsOpt?.LocalDeclarations[0].DeclaredTypeOpt; [](start = 12, length = 126)

If the only thing we need a type, consider getting it right here. Otherwise, it is not clear why we are looking only at the first local declaration

AlekseyTs replied:
>DeclaredTypeOpt [](start = 122, length = 15)

This looks fragile, can this be null. Perhaps it would be more robust to pull the type from the local symbol?

AlekseyTs commented on src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:
```diff
@@ -1906,7 +1909,8 @@ private IOperation CreateBoundMultipleLocalDeclarationsBaseOperation(BoundMultip
                     isAsynchronous: usingDecl.AwaitOpt is object,
                     usingDecl.DisposeMethodOpt.GetPublicSymbol(),
-                    _semanticModel,
+                    CreateDisposeArguments(usingDecl.DisposeMethodOpt, usingDecl.LocalDeclarations[0].DeclaredTypeOpt?.Type, usingDecl.Syntax, usingDecl.Binder),
```
>DeclaredTypeOpt [](start = 102, length = 15)

This looks fragile, can this be null? Perhaps it would be more robust to pull the type from the local symbol?

chsienki replied:
We don't have a local symbol for the using declaration. I've split it out and added an assert that shows we always have a type if we found a dispose method.

---
In reply to: [539463139](https://github.com/dotnet/roslyn/pull/49842#discussion_r539463139) [](ancestors = 539463139)

AlekseyTs replied:
> We don't have a local symbol for the using declaration.

What about the ```BoundLocalDeclaration.LocalSymbol``` property?

---
In reply to: [539652669](https://github.com/dotnet/roslyn/pull/49842#discussion_r539652669) [](ancestors = 539652669,539463139)

AlekseyTs commented on src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:
```diff
@@ -2337,5 +2341,44 @@ private IInstanceReferenceOperation CreateCollectionValuePlaceholderOperation(Bo
+
+        private ImmutableArray<IArgumentOperation> CreateDisposeArguments(MethodSymbol? disposeMethod, TypeSymbol? receiverType, SyntaxNode receiverSyntax, Binder binder)
+        {
+            if (disposeMethod is null || receiverType is null || disposeMethod.ParameterCount == 0)
```
>receiverType is null [](start = 41, length = 20)

The fact that we can come here without the type feels unexpected. Shouldn't we always know the type?

AlekseyTs commented on src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:
```diff
@@ -2337,5 +2341,44 @@ private IInstanceReferenceOperation CreateCollectionValuePlaceholderOperation(Bo
+
+            var argumentsBuilder = ArrayBuilder<BoundExpression>.GetInstance(disposeMethod.ParameterCount);
+            ImmutableArray<int> argsToParams = default;
+            var expanded = disposeMethod.HasParamsParameter();
```
>HasParamsParameter [](start = 41, length = 18)

This looks wrong. This helper doesn't perform the check consistent with OverloadResolution.IsValidParams. It also looks like all other usages have the same problem, possibly with exception of ExplicitInterfaceHelpers because they only affect diagnostics in a very restricted scenario. I think we should open an issue to clean all this up, but here we should do the right thing.

AlekseyTs replied:
Ideally we should capture this value during initial binding. I assume we performed overload resolution over available candidates, that should have given us the information about the best candidate and whether it should be used in an expanded form. We should not be trying to re-infer this aspect again.

---
In reply to: [539480603](https://github.com/dotnet/roslyn/pull/49842#discussion_r539480603) [](ancestors = 539480603)

AlekseyTs commented on src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:
```diff
@@ -2337,5 +2341,44 @@ private IInstanceReferenceOperation CreateCollectionValuePlaceholderOperation(Bo
+                argumentsBuilder,
+                argumentRefKindsBuilder: null,
+                ref argsToParams,
+                out var defaultArguments,
```
>var [](start = 20, length = 3)

How one is supposed to guess what type is this?

AlekseyTs commented on src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:
```diff
@@ -2337,5 +2341,44 @@ private IInstanceReferenceOperation CreateCollectionValuePlaceholderOperation(Bo
+                            argumentRefKindsOpt: default,
+                            expanded,
+                            receiver.Syntax,
+                            invokedAsExtensionMethod: false);
```
>invokedAsExtensionMethod: false [](start = 28, length = 31)

Consider asserting somewhere that the method is not static.

AlekseyTs commented on src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IUsingStatement.cs:
```diff
@@ -3131,6 +3131,363 @@ public Task DisposeAsync()
+            Statements (1)
+                IAwaitOperation (OperationKind.Await, Type: System.Void, IsInvalid, IsImplicit) (Syntax: 'this')
+                  Expression: 
+                    IInvocationOperation (virtual System.Threading.Tasks.ValueTask System.IAsyncDisposable.DisposeAsync()) (OperationKind.Invocation, Type: System.Threading.Tasks.ValueTask, IsInvalid, IsImplicit) (Syntax: 'this')
```
>IInvocationOperation (virtual System.Threading.Tasks.ValueTask System.IAsyncDisposable.DisposeAsync()) (OperationKind.Invocation, Type: System.Threading.Tasks.ValueTask, IsInvalid, IsImplicit) (Syntax: 'this') [](start = 20, length = 209)

This looks wrong. Only "valid" (amount of arguments correspond to parameters, etc.) invocations should be represented as IInvocationOperation. Also, it looks like System.IAsyncDisposable.DisposeAsync is not a valid candidate for pattern based dispose. We shouldn't treat it as one, The DisposeMethod should be null.

AlekseyTs commented on src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IUsingStatement.cs:
```diff
@@ -5855,7 +6212,7 @@ void M(bool b)
         Block[B3] - Block
             Predecessors (0)
             Statements (1)
-                IInvocationOperation (virtual void P.Dispose()) (OperationKind.Invocation, Type: System.Void, IsImplicit) (Syntax: 'x = new P()')
```
>virtual [](start = 38, length = 7)

Do we have at least one test with a pattern-based disposal utilizing a virtual method?

chsienki replied:
Added UsingDeclaration_Flow_23 and 24

---
In reply to: [539514783](https://github.com/dotnet/roslyn/pull/49842#discussion_r539514783) [](ancestors = 539514783)

AlekseyTs commented on src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IUsingStatement.cs:
```diff
@@ -3131,6 +3131,363 @@ public Task DisposeAsync()
+        await using(this){}
+    }/*</bind>*/
+    
+    Task DisposeAsync(int a = 3, params int[] extras, bool b = false) => throw null;
```
>   Task DisposeAsync(int a = 3, params int[] extras, bool b = false) => throw null; [](start = 1, length = 83)

I think we need to add more tests around an unconventional use of params/ParamArrayAttribute. For example:
1) What happens if we also add default value for ```extras``` parameter;
2) What happens if we consume a method like from #1 from metadata;
3) The last argument has params in metadata, but it is not an array and not optional; 
4) The last argument has params in metadata, but it is not an array, it is optional and doesn't have explicit default value; 
5) The last argument has params in metadata, but it is not an array, it is optional and has default value; 
6) The last argument has params in metadata, it is an array, it is optional and has default value.

We should also make sure that we have tests that are trying to emit code for scenarios like that, and verify what happens when emit succeeds.

AlekseyTs commented on src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IVariableDeclaration.cs:
```diff
@@ -2287,6 +2287,69 @@ static void Main(string[] args)
 
+        [CompilerTrait(CompilerFeature.IOperation)]
+        [Fact]
+        public void UsingStatement_DefaultDisposeArguments()
```
>UsingStatement_DefaultDisposeArguments [](start = 20, length = 38)

It feels like we should have some coverage for CFG as well.

chsienki replied:
See UsingFlow_18/19/20

---
In reply to: [539529809](https://github.com/dotnet/roslyn/pull/49842#discussion_r539529809) [](ancestors = 539529809)

AlekseyTs commented on src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IVariableDeclaration.cs:
```diff
@@ -2287,6 +2287,69 @@ static void Main(string[] args)
 
+        [CompilerTrait(CompilerFeature.IOperation)]
+        [Fact]
+        public void UsingStatement_DefaultDisposeArguments()
```
>UsingStatement_DefaultDisposeArguments [](start = 20, length = 38)

I am somewhat confused about placement of this test. It looks like it tests IUsingOperation node, but we have a dedicated test file for it.

AlekseyTs commented on src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IUsingStatement.cs:
```diff
@@ -6484,5 +6841,63 @@ public static async Task M1()
+{
+    public static void M1()
+    {
+        /*<bind>*/using var s = new S();/*</bind>*/
```
>using var s = new S(); [](start = 18, length = 22)

Do we have a CFG coverage for using declarations vs. using statements?

chsienki replied:
Added `UsingDeclaration_Flow_22`

---
In reply to: [539537373](https://github.com/dotnet/roslyn/pull/49842#discussion_r539537373) [](ancestors = 539537373)

AlekseyTs commented on src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IForEachLoopStatement.cs:
```diff
@@ -6031,7 +6031,7 @@ static void Main()
             Predecessors (0)
             Statements (1)
-                IInvocationOperation (virtual void C.Enumerator.Dispose()) (OperationKind.Invocation, Type: System.Void, IsImplicit) (Syntax: 'new C()')
+                IInvocationOperation ( void C.Enumerator.Dispose()) (OperationKind.Invocation, Type: System.Void, IsImplicit) (Syntax: 'new C()')
```
It doesn't look like any CFG coverage added for arguments with `foreach`

chsienki replied:
Added `Foreach_RefStructEnumerator_DefaultDisposeArguments`

---
In reply to: [539538730](https://github.com/dotnet/roslyn/pull/49842#discussion_r539538730) [](ancestors = 539538730)

AlekseyTs commented:
Done with review pass (commit 3)

chsienki commented:
Thanks again @AlekseyTs I think I've addressed all the feedback, with the exception of the params array which i'll work on next.

AlekseyTs commented on src/Compilers/Core/Portable/Operations/DisposeOperationInfo.cs:
```diff
@@ -0,0 +1,21 @@
+{
+    internal struct DisposeOperationInfo
+    {
+        public readonly IMethodSymbol DisposeMethod;
```
>IMethodSymbol [](start = 24, length = 13)

Nullable?

AlekseyTs commented on src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:
```diff
@@ -2337,5 +2349,48 @@ private IInstanceReferenceOperation CreateCollectionValuePlaceholderOperation(Bo
+                new DiagnosticBag());
+
+            var args = DeriveArguments(
+                            receiver,
```
>receiver [](start = 28, length = 8)

It looks like this parameter is not used (at least in master). If that is the case, I think we should remove it and simplify consumers. For example, for this change we wouldn't need to create the receiver node and, therefore, wouldn't need to go through the trouble of figuring out its type in order to call ```CreateDisposeArguments```.

chsienki replied:
Oh weird, good catch. There are a couple of unused arguments to that method. I've removed them and simplified the call sites to not need the type.

---
In reply to: [540155631](https://github.com/dotnet/roslyn/pull/49842#discussion_r540155631) [](ancestors = 540155631)

AlekseyTs commented:
Done with review pass (commit 13)

AlekseyTs commented on src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:
```diff
@@ -2337,5 +2346,45 @@ private IInstanceReferenceOperation CreateCollectionValuePlaceholderOperation(Bo
+            ImmutableArray<int> argsToParams = default;
+            var expanded = disposeMethod.HasParamsParameter();
+
+            Debug.Assert(!expanded || disposeMethod.GetParameters().Last().Type.IsArray());
```
>IsArray [](start = 80, length = 7)

Not just any array, an array of specific shape, see ```OverloadResolution.IsValidParams```

AlekseyTs commented on src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory_Methods.cs:
```diff
@@ -200,14 +200,11 @@ internal ImmutableArray<IArgumentOperation> DeriveArguments(BoundNode containing
                                     boundObjectInitializerMember.Binder,
                                     property,
                                     boundObjectInitializerMember.Arguments,
-                                    boundObjectInitializerMember.ArgumentNamesOpt,
```
>ArgumentNamesOpt [](start = 65, length = 16)

Is this property used anywhere else? Can it be removed?

chsienki replied:
Yes, these and all the others are used in various places such as lowering and diagnostic passes where they are needed to visit the call.

---
In reply to: [541219591](https://github.com/dotnet/roslyn/pull/49842#discussion_r541219591) [](ancestors = 541219591)

AlekseyTs commented on src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory_Methods.cs:
```diff
@@ -200,14 +200,11 @@ internal ImmutableArray<IArgumentOperation> DeriveArguments(BoundNode containing
-                                    boundObjectInitializerMember.ArgumentNamesOpt,
                                     boundObjectInitializerMember.ArgsToParamsOpt,
                                     boundObjectInitializerMember.DefaultArguments,
-                                    boundObjectInitializerMember.ArgumentRefKindsOpt,
```
>ArgumentRefKindsOpt [](start = 65, length = 19)

Is this property used anywhere else? Can it be removed?

AlekseyTs commented on src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory_Methods.cs:
```diff
@@ -225,44 +222,35 @@ internal ImmutableArray<IArgumentOperation> DeriveArguments(BoundNode containing
+                        return DeriveArguments(boundIndexer.BinderOpt,
                                                boundIndexer.Indexer,
                                                boundIndexer.Arguments,
-                                               boundIndexer.ArgumentNamesOpt,
```
>ArgumentNamesOpt [](start = 60, length = 16)

Is this property used anywhere else? Can it be removed?

AlekseyTs commented on src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory_Methods.cs:
```diff
@@ -225,44 +222,35 @@ internal ImmutableArray<IArgumentOperation> DeriveArguments(BoundNode containing
-                                               boundIndexer.ArgumentNamesOpt,
                                                boundIndexer.ArgsToParamsOpt,
                                                boundIndexer.DefaultArguments,
-                                               boundIndexer.ArgumentRefKindsOpt,
```
>ArgumentRefKindsOpt [](start = 60, length = 19)

Is this property used anywhere else? Can it be removed?

AlekseyTs commented on src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory_Methods.cs:
```diff
@@ -225,44 +222,35 @@ internal ImmutableArray<IArgumentOperation> DeriveArguments(BoundNode containing
+                        return DeriveArguments(objectCreation.BinderOpt,
                                                objectCreation.Constructor,
                                                objectCreation.Arguments,
-                                               objectCreation.ArgumentNamesOpt,
```
>ArgumentNamesOpt [](start = 62, length = 16)

Is this property used anywhere else? Can it be removed?

AlekseyTs commented on src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory_Methods.cs:
```diff
@@ -225,44 +222,35 @@ internal ImmutableArray<IArgumentOperation> DeriveArguments(BoundNode containing
-                                               objectCreation.ArgumentNamesOpt,
                                                objectCreation.ArgsToParamsOpt,
                                                objectCreation.DefaultArguments,
-                                               objectCreation.ArgumentRefKindsOpt,
```
>ArgumentRefKindsOpt [](start = 62, length = 19)

Is this property used anywhere else? Can it be removed?

AlekseyTs commented on src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory_Methods.cs:
```diff
@@ -225,44 +222,35 @@ internal ImmutableArray<IArgumentOperation> DeriveArguments(BoundNode containing
+                        return DeriveArguments(boundCall.BinderOpt,
                                                boundCall.Method,
                                                boundCall.Arguments,
-                                               boundCall.ArgumentNamesOpt,
```
>ArgumentNamesOpt [](start = 57, length = 16)

Is this property used anywhere else? Can it be removed?

AlekseyTs commented on src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory_Methods.cs:
```diff
@@ -225,44 +222,35 @@ internal ImmutableArray<IArgumentOperation> DeriveArguments(BoundNode containing
-                                               boundCall.ArgumentNamesOpt,
                                                boundCall.ArgsToParamsOpt,
                                                boundCall.DefaultArguments,
-                                               boundCall.ArgumentRefKindsOpt,
```
>ArgumentRefKindsOpt [](start = 57, length = 19)

Is this property used anywhere else? Can it be removed?

AlekseyTs reviewed (approved):
LGTM (commit 18), assuming CI is passing.

333fred commented on src/Compilers/Core/Portable/Operations/DisposeOperationInfo.cs:
```diff
@@ -0,0 +1,21 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+#nullable enable
```
This should be the default unless the file is generated.

333fred commented on src/Compilers/Core/Portable/Operations/DisposeOperationInfo.cs:
```diff
@@ -0,0 +1,21 @@
+
+namespace Microsoft.CodeAnalysis.Operations
+{
+    internal struct DisposeOperationInfo
```
Feels like this file should be located in the same place as `ForEachEnumeratorInfo`.

333fred commented:
You'll need to rerun the generator locally to pick up the children changes I made in the main branch.

333fred reviewed (approved):
LGTM (commit 20). Just a couple of minor formatting changes.

chsienki commented:
Rebasing to target master
